File: examples/minimart-go-nuxt/backend/internal/application/usecase/product_usecase.go
package usecase

import (
	"context"

	"minimart/internal/application/dto"
	"minimart/internal/infrastructure/repository"
)

// ProductUsecase reads the public catalogue.
type ProductUsecase struct {
	products *repository.ProductRepository
}

// NewProductUsecase returns a ProductUsecase over products.
func NewProductUsecase(products *repository.ProductRepository) *ProductUsecase {
	return &ProductUsecase{products: products}
}

// List returns the active catalogue. An empty catalogue is an empty list, not
// an error, and Items is always a JSON array — never null.
func (uc *ProductUsecase) List(ctx context.Context) (*dto.ProductList, error) {
	products, err := uc.products.ListActive(ctx)
	if err != nil {
		return nil, err
	}

	items := make([]dto.Product, 0, len(products))
	for _, p := range products {
		items = append(items, dto.Product{
			ID:         p.ID,
			Name:       p.Name,
			CostPoints: p.CostPoints,
			Stock:      p.Stock,
		})
	}
	return &dto.ProductList{Items: items, Total: len(items)}, nil
}

File: examples/minimart-java-next/backend/src/main/java/minimart/application/usecase/ProductUsecase.java
package minimart.application.usecase;

import minimart.application.dto.ProductItem;
import minimart.application.dto.ProductList;
import minimart.infrastructure.repository.ProductRepository;

/** Reads the public catalogue. */
public class ProductUsecase {

  private final ProductRepository products;

  /** A use case over {@code products}. */
  public ProductUsecase(ProductRepository products) {
    this.products = products;
  }

  /**
   * Returns the active catalogue. An empty catalogue is an empty list, not an error, and {@code
   * items} is always a JSON array — never null.
   */
  public ProductList list() {
    return ProductList.of(
        products.listActive().stream()
            .map(p -> new ProductItem(p.id(), p.name(), p.costPoints(), p.stock()))
            .toList());
  }
}
